iModelCore/ECDb/Tests/NonPublished/InstanceReaderTest.cpp

```diff
@@ -1769,6 +1769,87 @@ TEST_F(InstanceReaderFixture, dynamic_meta_data) {
+                <ECProperty propertyName='prop' typeName='int' />
+            </ECEntityClass>
+            <ECEntityClass typeName='C'>
+                <BaseClass>A</BaseClass>
+                <ECProperty propertyName='prop' typeName='double' />
+            </ECEntityClass>
+        </ECSchema>)xml")));
+
+    auto exec = [&](Utf8CP ecsql) {
+        ECSqlStatement stmt;
+        EXPECT_EQ(ECSqlStatus::Success, stmt.Prepare(m_ecdb, ecsql));
+        EXPECT_EQ(BE_SQLITE_DONE, stmt.Step());
+        m_ecdb.SaveChanges();
+    };
+
+    auto assertDefault = [](ECSqlStatement& stmt, int cl, Utf8CP displayLabel, Utf8CP propertyName) {
+        ECSqlColumnInfo const* ci;
+        PrimitiveECPropertyCP pr;
+        Utf8CP className = "DynamicECSqlSelectClause";
+        ci = &stmt.GetColumnInfo(cl);
+        EXPECT_TRUE(ci->IsDynamic());
+        EXPECT_TRUE(ci->GetDataType().IsPrimitive());
+        pr = ci->GetProperty()->GetAsPrimitiveProperty();
+        EXPECT_FALSE(pr->HasId());
+        EXPECT_EQ(PrimitiveType::PRIMITIVETYPE_String, pr->GetType());
+        EXPECT_STRCASEEQ(className                   , pr->GetClass().GetName().c_str());
+        EXPECT_STRCASEEQ(propertyName                , pr->GetName().c_str());
+        EXPECT_STRCASEEQ(""                          , pr->GetDescription().c_str());
+        EXPECT_STRCASEEQ(displayLabel                , pr->GetDisplayLabel().c_str());
+        EXPECT_STRCASEEQ("json"                      , pr->GetExtendedTypeName().c_str());
+    };
+
+    auto assertDynamic = [](ECSqlStatement& stmt, int cl, Utf8CP displayLabel, Utf8CP propertyName, Utf8CP description, Utf8CP className, PrimitiveType t) {
+        ECSqlColumnInfo const* ci;
+        PrimitiveECPropertyCP pr;
+        ci = &stmt.GetColumnInfo(cl);
+        EXPECT_TRUE(ci->IsDynamic());
+        EXPECT_TRUE(ci->GetDataType().IsPrimitive());
+        pr = ci->GetProperty()->GetAsPrimitiveProperty();
+        EXPECT_TRUE(pr->HasId());
+        EXPECT_EQ(t                  , pr->GetType());
+        EXPECT_STRCASEEQ(className   , pr->GetClass().GetName().c_str());
+        EXPECT_STRCASEEQ(propertyName, pr->GetName().c_str());
+        EXPECT_STRCASEEQ(description , pr->GetDescription().c_str());
+        EXPECT_STRCASEEQ(displayLabel, pr->GetDisplayLabel().c_str());
+        EXPECT_STRCASEEQ(""          , pr->GetExtendedTypeName().c_str());
+    };
+
+    exec("insert into ts.B ( ecInstanceId, B.prop ) values ( 10, 100 )");
+    exec("insert into ts.C ( ecInstanceId, C.prop ) values ( 11, 101 )");
+
+    if ("top level instance props meta data") {
+        const auto sql = R"x(
+            select
+                $->prop
+            from ts.A
+        )x";
+
+        ECSqlStatement stmt;
+        ASSERT_EQ(ECSqlStatus::Success, stmt.Prepare(m_ecdb, sql));
+        assertDefault(stmt, 0 , "$->prop", "__x0024____x002D____x003E__prop");
+        while(stmt.Step() == BE_SQLITE_ROW) {
+            auto ecInstanceId = stmt.GetValueInt(0);
```

Comment:
`stmt.GetValueInt(0) is $->prop` not `ECInstanceId` which is not selected. The loop runs but I bet none of `assertDynamic()` ever executed. `ecInstanceId` be likely `0`. 

You can just do 

```
ASSERT_EQ(BE_SQLITE_ROW, stmt.Step())
assertDynamic(stmt, 0 , "prop", "prop", "", "B", PrimitiveType::PRIMITIVETYPE_Integer);
ASSERT_EQ(BE_SQLITE_ROW, stmt.Step())
assertDynamic(stmt, 0 , "prop", "prop", "", "C", PrimitiveType::PRIMITIVETYPE_Double);
```

Reply:
I have modified the test according to your comments. I missed this part when I removed `ecInstanceId` from the select statement. Thanks for pointing it out.